File: shopkeeper-angular/src/app/core/services/admin.service.ts
import { Injectable, inject } from '@angular/core';
import {
  Firestore, collection, collectionData, doc, setDoc, deleteDoc,
  getDocs, query, orderBy, serverTimestamp, CollectionReference,
} from '@angular/fire/firestore';
import { Observable } from 'rxjs';

import { UserProfile, Role, Plan } from '../models';

/**
 * Admin operations over the `users/*` collection. Every write here is also
 * authorised by the Firestore rules (admin-only), so the UI only exposes what
 * an admin is already permitted to do.
 */
@Injectable({ providedIn: 'root' })
export class AdminService {
  private readonly fs = inject(Firestore);
  private readonly businessCols = ['parties', 'items', 'txns', 'adjustments'];

  /** Live list of all registered users, newest first. */
  users$(): Observable<UserProfile[]> {
    const ref = collection(this.fs, 'users') as CollectionReference<UserProfile>;
    return collectionData<UserProfile>(query(ref, orderBy('createdAt', 'desc')));
  }

  setActive(uid: string, active: boolean): Promise<void> {
    const patch: Record<string, unknown> = { active };
    if (active) patch['activatedAt'] = serverTimestamp();
    return setDoc(doc(this.fs, 'users', uid), patch, { merge: true });
  }

  setRole(uid: string, role: Role): Promise<void> {
    return setDoc(doc(this.fs, 'users', uid), { role }, { merge: true });
  }

  setPlan(uid: string, plan: Plan): Promise<void> {
    return setDoc(doc(this.fs, 'users', uid), { plan, planUpdatedAt: serverTimestamp() }, { merge: true });
  }

  /** Remove the profile and the user's whole business-data subtree. */
  async deleteUser(uid: string): Promise<void> {
    for (const col of this.businessCols) {
      const snap = await getDocs(collection(this.fs, `userData/${uid}/${col}`));
      await Promise.all(snap.docs.map((d) => deleteDoc(d.ref)));
    }
    const meta = await getDocs(collection(this.fs, `userData/${uid}/meta`));
    await Promise.all(meta.docs.map((d) => deleteDoc(d.ref)));
    await deleteDoc(doc(this.fs, 'users', uid));
  }
}

File: shopkeeper-flutter/lib/core/services/admin_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';

import '../models/models.dart';

/// Admin operations over `users/*`. Every write is also authorised by the
/// Firestore rules (admin-only), so the UI only exposes permitted actions.
class AdminService {
  final FirebaseFirestore _db = FirebaseFirestore.instance;

  Stream<List<UserProfile>> users() => _db
      .collection('users')
      .orderBy('createdAt', descending: true)
      .snapshots()
      .map((s) => s.docs.map((d) => UserProfile.fromMap(d.id, d.data())).toList());

  Future<void> setActive(String uid, bool active) => _db.collection('users').doc(uid).set({
        'active': active,
        if (active) 'activatedAt': FieldValue.serverTimestamp(),
      }, SetOptions(merge: true));

  Future<void> setRole(String uid, String role) =>
      _db.collection('users').doc(uid).set({'role': role}, SetOptions(merge: true));

  Future<void> setPlan(String uid, String plan) => _db.collection('users').doc(uid).set(
      {'plan': plan, 'planUpdatedAt': FieldValue.serverTimestamp()}, SetOptions(merge: true));

  Future<void> deleteUser(String uid) async {
    for (final col in ['parties', 'items', 'txns', 'adjustments']) {
      final snap = await _db.collection('userData').doc(uid).collection(col).get();
      for (final d in snap.docs) {
        await d.reference.delete();
      }
    }
    final meta = await _db.collection('userData').doc(uid).collection('meta').get();
    for (final d in meta.docs) {
      await d.reference.delete();
    }
    await _db.collection('users').doc(uid).delete();
  }
}
